src/Data/Text/Internal/Validate.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+
+-- | Test whether or not a sequence of bytes is a valid UTF-8 byte sequence.
+-- In the GHC Haskell ecosystem, there are several representations of byte
+-- sequences. The only one that the stable @text@ API concerns itself with is
+-- 'ByteString'. Part of bytestring-to-text decoding is 'isValidUtf8ByteString',
+-- a high-performance UTF-8 validation routine written in C++ with fallbacks
+-- for various platforms. The C++ code backing this routine is nontrivial,
+-- so in the interest of reuse, this module additionally exports functions
+-- for working with the GC-managed @ByteArray#@ type. These @ByteArray#@
+-- functions are not used anywhere else in @text@. They are for the benefit
+-- of library and application authors who do not use 'ByteString' but still
+-- need to interoperate with @text@.
+--
+-- Note: The release of GHC 9.4 has seen a coordinated move of the lifted
+-- @ByteArray@ type from @primitive@ to @base@. In many years, when support
+-- for GHC older than 9.4 is dropped, it will be possible to add functions
+-- for the lifted ByteArray type as well.
```

Comment:
Please rebase your branch atop of https://github.com/haskell/text/pull/474 and switch types to a lifted `type Array = ByteArray` from `Data.Text.Array`. I'm a bit reluctant to introduce API for unlifted arrays just to be painfully deprecated (or extended) soon.

Reply:
I have no strong preferences here.